python-package/lightgbm/dask.py

```diff
@@ -282,17 +354,100 @@ def _train(
     data_parts = _split_to_parts(data=data, is_matrix=True)
     label_parts = _split_to_parts(data=label, is_matrix=False)
     parts = [{'data': x, 'label': y} for (x, y) in zip(data_parts, label_parts)]
+    n_parts = len(parts)
 
     if sample_weight is not None:
         weight_parts = _split_to_parts(data=sample_weight, is_matrix=False)
-        for i in range(len(parts)):
+        for i in range(n_parts):
             parts[i]['weight'] = weight_parts[i]
 
     if group is not None:
         group_parts = _split_to_parts(data=group, is_matrix=False)
-        for i in range(len(parts)):
+        for i in range(n_parts):
             parts[i]['group'] = group_parts[i]
 
+    # evals_set will to be re-constructed into smaller lists of (X, y) tuples, where
+    # X and y are each delayed sub-lists of original eval dask Collections.
+    if eval_set:
+        eval_sets = defaultdict(list)
+        if eval_sample_weight:
+            eval_sample_weights = defaultdict(list)
+        if eval_group:
+            eval_groups = defaultdict(list)
+
+        for i, (X, y) in enumerate(eval_set):
+
+            # when individual eval set is equivalent to training data, skip recomputing parts.
+            if id(X) == id(data) and id(y) == id(label):
```

Comment:
```suggestion
            if X is data and y is label
```

Is there a particular reason to prefer `id()` to using `is`? I honestly don't know if there's any meaningful difference.

For this check in `sklearn` we use `is`, so if you don't have a strong preference I'd prefer to change to `is` for consistency and because I think people who might come contribute are more likely to be familiar with `is` than they are with `id()`.

https://github.com/microsoft/LightGBM/blob/fd094d2719a04b984b376a665b6c3e5e1aff2967/python-package/lightgbm/sklearn.py#L663

Reply:
Yeah I just took this from xgboost_dask.py - https://github.com/dmlc/xgboost/blob/master/python-package/xgboost/dask.py#L865

I mean `is` is easier to read, let's do it.